client/consensus/babe/src/lib.rs

```diff
@@ -1542,6 +1543,35 @@ where
 					log::Level::Info
 				};
 
+				if viable_epoch.as_ref().end_slot() <= slot {
+					// some epochs must have been skipped as our current slot
+					// fits outside the current epoch. we will figure out
+					// which epoch it belongs to and we will re-use the same
+					// data for that epoch
+					let mut epoch_data = viable_epoch.as_mut();
+					let skipped_epochs = (*slot - *epoch_data.start_slot) / epoch_data.duration;
+
+					let original_epoch_index = epoch_data.epoch_index;
+
+					// NOTE: notice that we are only updating a local copy of the `Epoch`, this
+					// makes it so that when we insert the next epoch into `EpochChanges` below
+					// (after incrementing it), it will use the correct epoch index and start slot.
+					// we do not update the original epoch that will be re-used because there might
+					// be other forks (that we haven't imported) where the epoch isn't skipped, and
+					// to import those forks we want to keep the original epoch data. not updating
+					// the original epoch works because when we search the tree for which epoch to
+					// use for a given slot, we will search in-depth with the predicate
+					// `epoch.start_slot <= slot` which will still match correctly without updating
+					// `start_slot` to the correct value as below.
```

Comment:
This comment talks about skipping  an epoch, and later trying to import blocks that were actually authored in that skipped epoch that we previously did not import, correct?
Should we have a test for this scenario?

Reply:
Good point. Done 👍